src/3d/qgstessellator.cpp

```diff
@@ -157,68 +182,160 @@ void QgsTessellator::addPolygon( const QgsPolygonV2 &polygon, float extrusionHei
   polyline.reserve( exterior->numPoints() );
 
   QgsVertexId::VertexType vt;
-  QgsPoint pt, ptPrev;
+  QgsPoint pt;
 
-  for ( int i = 0; i < exterior->numPoints() - 1; ++i )
+  QVector3D pNormal = _calculateNormal( exterior );
+  const int pCount = exterior->numPoints();
+
+  // Polygon is a triangle
+  if ( pCount == 4 )
   {
-    exterior->pointAt( i, pt, vt );
-    if ( i == 0 || pt != ptPrev )
+    QgsPoint pt;
+    for ( int i = 0; i < 3; i++ )
     {
-      p2t::Point *pt2 = new p2t::Point( pt.x() - mOriginX, pt.y() - mOriginY );
-      polyline.push_back( pt2 );
-      float zPt = qIsNaN( pt.z() ) ? 0 : pt.z();
-      z[pt2] = zPt;
+      exterior->pointAt( i, pt, vt );
+      mData << pt.x() - mOriginX << pt.z() << - pt.y() + mOriginY;
+      if ( mAddNormals )
+        mData << pNormal.x() << pNormal.z() << - pNormal.y();
     }
-    ptPrev = pt;
   }
-  polylinesToDelete << polyline;
+  else
+  {
 
-  p2t::CDT *cdt = new p2t::CDT( polyline );
+    if ( !qgsDoubleNear( pNormal.length(), 1, 0.001 ) )
+    {
+      return;
+    }
 
-  // polygon holes
-  for ( int i = 0; i < polygon.numInteriorRings(); ++i )
-  {
-    std::vector<p2t::Point *> holePolyline;
-    holePolyline.reserve( exterior->numPoints() );
-    const QgsCurve *hole = polygon.interiorRing( i );
-    for ( int j = 0; j < hole->numPoints() - 1; ++j )
+    const QgsPoint ptFirst( exterior->startPoint() );
+    QVector3D pOrigin( ptFirst.x(), ptFirst.y(), ptFirst.z() ), pXVector;
+    // Here we define the two perpendicular vectors that define the local
+    // 2D space on the plane. They will act as axis for which we will
+    // calculate the projection coordinates of a 3D point to the plane.
+    if ( pNormal.z() > 0.001 || pNormal.z() < -0.001 )
+    {
+      pXVector = QVector3D( 1, 0, -pNormal.x() / pNormal.z() );
+    }
+    else if ( pNormal.y() > 0.001 || pNormal.y() < -0.001 )
+    {
+      pXVector = QVector3D( 1, -pNormal.x() / pNormal.y(), 0 );
+    }
+    else
     {
-      hole->pointAt( j, pt, vt );
-      if ( j == 0 || pt != ptPrev )
+      pXVector = QVector3D( -pNormal.y() / pNormal.x(), 1, 0 );
+    }
+    pXVector.normalize();
+    QVector3D pYVector = QVector3D::normal( pNormal, pXVector );
```

Comment:
pYVector const 

As a side note, triadic operator instead of `if` would allow for const pYVector:
```
const pXvector(
  abs(pNormal.z() > 0.001 
  ? QVector3D( 1, 0, -pNormal.x() / pNormal.z() ).normalized()
  :  
    pNormal.y() > 0.001
    ? QVector3D( 1, -pNormal.x() / pNormal.y(), 0 ).normalized() 
    :  QVector3D( -pNormal.y() / pNormal.x(), 1, 0 ).normalized()
);
```
but I must admit it's a bit less readable with 3 cases.